### PR TITLE
feat: add DID pallet benchmarks

### PR DESCRIPTION
**Pull request by Diiaablo95**
This PR adds benchmarking for the DID pallet, and three more parameters in the pallet configuration: one for the maximum length of endpoint URLs, one for the maximum number of new key agreement keys in a create or update operation, and one for the maximum number of public keys that can be removed in an update operation.

- fixes KILTProtocol/ticket#1197
- fixes KILTProtocol/ticket#1362

## How to test:

Run `cargo test --all --all-features`.

### Custom types

This PR does not introduce new types.

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

**Comment by wischli** [comment]
Sorry, won't be able to do reviews until the staking pallet is ready for the security audit.

**Review by weichweich — CHANGES_REQUESTED** [review]
the submit did call also needs benchmarking or at least some weights.

The weights should be the the base weight plus the weight of the inner call. Maybe have a look at the proxy pallet. They had a similar situation.

In the benchmarks they used 
```let call: <T as Config>::Call = frame_system::Call::<T>::remark(vec![]).into();```
as a call and 
```
		#[pallet::weight({
			let di = call.get_dispatch_info();
			(T::WeightInfo::proxy(T::MaxProxies::get().into())
				.saturating_add(di.weight)
				 // AccountData for inner call origin accountdata.
				.saturating_add(T::DbWeight::get().reads_writes(1, 1)),
			di.class)
		})]
```
as weight.

**Comment by Diiaablo95** [comment]
@weichweich I guess you could now take another look at the PR, perhaps starting from commit https://github.com/KILTprotocol/mashnet-node/pull/162/commits/2c7d58e8a1c77d16178175c136f713eda5efba52.

Beyond the required change in the use of weights for the dispatch call operation, I had to introduce the `#[cfg(feature = "runtime-benchmarks")] feature check in a couple more places, otherwise the DID pallet would not compile or the CTYPE pallet would throw `BadOrigin` errors when testing the DID pallet for dispatch calls. I think we should be good to go now! 🏍🏍🏍

File: pallets/did/src/benchmarking.rs [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+		);
+		assert_eq!(stored_did.endpoint_url, did_creation_op.new_endpoint_url);
+		assert_eq!(stored_did.last_tx_counter, 0u64);
+	}
+
+	submit_did_update_operation_ed25519_keys {
+		let n in 1 .. T::MaxNewKeyAgreementKeys::get();
+		let m in 1 .. T::MaxVerificationKeysToRevoke::get();
+		// Doesn't allow me to reference the const declared above
+		let u in 9 .. T::MaxUrlLength::get();
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/benchmarking.rs]: Fixed in https://github.com/KILTprotocol/mashnet-node/pull/162/commits/61e177415f2a463f6a04a5fb594aaafe2f1b1816

File: pallets/did/src/benchmarking.rs [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+		let did_subject = DidIdentifierOf::<T>::default();
+		let did_public_auth_key = get_ed25519_public_authentication_key();
+
+		let did_details = get_did_base_details(DidVerificationKey::from(did_public_auth_key));
+		Did::<T>::insert(&did_subject, did_details);
+
+		let did_deletion_op = generate_base_did_deletion_operation::<T>(did_subject.clone());
+
+		let did_deletion_signature = ed25519_sign(AUTHENTICATION_KEY_ID, &did_public_auth_key, did_deletion_op.encode().as_ref()).expect("Failed to create DID signature from raw ed25519 signature.");
+	}: submit_did_delete_operation(RawOrigin::Signed(submitter), did_deletion_op.clone(), DidSignature::from(did_deletion_signature))
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/benchmarking.rs]: See comment below.

File: pallets/did/src/lib.rs [hunk 3]
```diff
@@ -81,9 +91,19 @@ pub mod pallet {
 			+ Dispatchable<Origin = <Self as Config>::Origin, PostInfo = PostDispatchInfo>
 			+ GetDispatchInfo
 			+ DeriveDidCallAuthorizationVerificationKeyRelationship;
-		type DidIdentifier: Parameter;
+		type DidIdentifier: Parameter + Default;
+		#[cfg(not(feature = "runtime-benchmarks"))]
 		type Origin: From<DidRawOrigin<DidIdentifierOf<Self>>>;
+		#[cfg(feature = "runtime-benchmarks")]
+		type Origin: From<RawOrigin<DidIdentifierOf<Self>>>;
```
Comment (weichweich) [inline pallets/did/src/lib.rs:98]: Mh that's strange. I would like to not have this check here. Maybe we can do some rust magic.

**Review by weichweich — COMMENTED** [review]
looks good. But i would like to investigate this origin thing.

Also merging is blocked until we have a new mashnet release. 😓

File: pallets/did/src/lib.rs (hunk 3 above)
Comment (Diiaablo95) [reply pallets/did/src/lib.rs:98]: Mhh yeah I agree, but the whole benchmarking thing has been somehow "hardcoded" towards using specific functions and keys to "make it work". But hopefully there is a better method.

**Review by weichweich — APPROVED** [review]
LGTM, as discussed we need to think a bit more about the origin thing.

**Pull request merged**